README.md
# bip_bech32

[![pub package](https://img.shields.io/pub/v/bip_bech32.svg)](https://pub.dartlang.org/packages/bip_bech32)

Dart implementing [BIP173 spec](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki)

Convert from [bitcoinj](https://github.com/bitcoinj/bitcoinj/blob/master/core/src/main/java/org/bitcoinj/core/Bech32.java)

## Install

dart pub package

[bip_bech32](https://pub.dartlang.org/packages/bip_bech32)

## Example

```dart
Bech32Codec bech32codec = Bech32Codec();
  Bech32 bech32 =
      bech32codec.decode("bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4");
  print(bech32.data);
  print(bech32.hrp);

  Uint8List data = Uint8List.fromList([0, 14, 20, 15, 7, 13, 26, 0, 25, 18, 6, 11, 13, 8, 21, 4, 20, 3, 17, 2, 29, 3, 12, 29, 3, 4, 15, 24, 20, 6, 14, 30, 22]);
  print(bech32codec.encode(Bech32("bc", data)));
```

[example](https://github.com/OrangeWallet/bech32/blob/master/example/main.dart)

## License

[MIT](https://github.com/OrangeWallet/bech32/blob/master/LICENSE.md)

## Thanks

[bitcoinj](https://github.com/bitcoinj/bitcoinj)


CHANGELOG.md
# CHANGELOG

## 0.1.3

- fix library import

## 0.1.2

- Initial release


test/bech32_test.dart
import 'dart:typed_data';

import 'package:bip_bech32/src/bech32.dart';
import 'package:test/test.dart';

main() {
  Uint8List data = Uint8List.fromList([
    0,
    14,
    20,
    15,
    7,
    13,
    26,
    0,
    25,
    18,
    6,
    11,
    13,
    8,
    21,
    4,
    20,
    3,
    17,
    2,
    29,
    3,
    12,
    29,
    3,
    4,
    15,
    24,
    20,
    6,
    14,
    30,
    22
  ]);

  test("bech32 encode", () {
    Bech32Codec bech32codec = Bech32Codec();
    expect("bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4",
        bech32codec.encode(Bech32("bc", data)));
  });

  test("bech32 decode", () {
    Bech32Codec bech32codec = Bech32Codec();
    expect(data,
        bech32codec.decode("bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4").data);
  });
}


lib/src/bech32.dart
import 'dart:convert';

import 'dart:typed_data';

import 'package:bip_bech32/src/core.dart';
import 'package:bip_bech32/src/exceptions.dart';

class Bech32 {
  Bech32(this.hrp, this.data);

  final String hrp;
  final Uint8List data;
}

class Bech32Codec extends Codec<Bech32, String> {
  Bech32Decoder get decoder => Bech32Decoder();
  Bech32Encoder get encoder => Bech32Encoder();

  String encode(Bech32 data) {
    return Bech32Encoder().convert(data);
  }

  Bech32 decode(String data) {
    return Bech32Decoder().convert(data);
  }
}

class Bech32Encoder extends Converter<Bech32, String> {
  @override
  String convert(Bech32 input) {
    String hrp = input.hrp;
    Uint8List data = input.data;
    if (hrp.isEmpty) throw TooShortHumanReadable(hrp.length);
    if (input.hrp.length > 83) throw TooLongHumanReadable(hrp.length);
    hrp = hrp.toLowerCase();
    Uint8List checksummed =
        Uint8List.fromList(data + createChecksum(hrp, data));
    return hrp + separator + checksummed.map((i) => charset[i]).join();
  }
}

class Bech32Decoder extends Converter<String, Bech32> {
  @override
  Bech32 convert(String str) {
    bool lower = false, upper = false;
    if (str.length < 8) throw TooShortInput(str.length);
    if (str.length > 90) throw TooLongInput(str.length);
    for (int i = 0; i < str.length; ++i) {
      int c = str.codeUnitAt(i);
      if (c < 33 || c > 126) throw InvalidCharacter(c, i);
      if (c >= 'a'.codeUnitAt(0) && c <= 'z'.codeUnitAt(0)) {
        if (upper) throw InvalidCharacter(c, i);
        lower = true;
      }
      if (c >= 'A'.codeUnitAt(0) && c <= 'Z'.codeUnitAt(0)) {
        if (lower) throw InvalidCharacter(c, i);
        upper = true;
      }
    }
    final int pos = str.lastIndexOf('1');
    if (pos < 1) throw MissingHumanPart();
    final int dataPartLength = str.length - 1 - pos;
    if (dataPartLength < 6) throw TooShortDataPart(dataPartLength);
    var values = Uint8List(dataPartLength);
    for (int i = 0; i < dataPartLength; ++i) {
      int c = str.codeUnitAt(i + pos + 1);
      if (CHARSET_REV[c] == -1) throw InvalidCharacter(c, i);
      values[i] = CHARSET_REV[c];
    }
    String hrp = str.substring(0, pos).toLowerCase();
    if (!verifyChecksum(hrp, values)) throw InvalidChecksum;
    return Bech32(hrp, copyOfRange(values, 0, values.length - 6));
  }
}


lib/src/exceptions.dart
class TooShortHumanReadable implements Exception {
  final int length;

  TooShortHumanReadable(this.length);

  String toString() => "Human-readable part is too short: $length";
}

class TooLongHumanReadable implements Exception {
  final int length;

  TooLongHumanReadable(this.length);

  String toString() => "Human-readable part is too long: $length";
}

class TooShortInput implements Exception {
  final int length;

  TooShortInput(this.length);

  String toString() => "Input part is too short: $length";
}

class TooLongInput implements Exception {
  final int length;

  TooLongInput(this.length);

  String toString() => "Input part is too long: $length";
}

class InvalidCharacter implements Exception {
  final int character;
  final int position;

  InvalidCharacter(this.character, this.position);

  String toString() => "Invalid character $character at position $position";
}

class InvalidChecksum implements Exception {
  String toString() => "Checksum does not validate";
}

class MissingHumanPart implements Exception {
  String toString() => "Missing human-readable part";
}

class TooShortDataPart implements Exception {
  final int length;

  TooShortDataPart(this.length);

  String toString() => "Data part too short: $length";
}


lib/src/core.dart
import 'dart:math';

import 'dart:typed_data';

const String separator = "1";

const charset = [
  "q",
  "p",
  "z",
  "r",
  "y",
  "9",
  "x",
  "8",
  "g",
  "f",
  "2",
  "t",
  "v",
  "d",
  "w",
  "0",
  "s",
  "3",
  "j",
  "n",
  "5",
  "4",
  "k",
  "h",
  "c",
  "e",
  "6",
  "m",
  "u",
  "a",
  "7",
  "l",
];

const CHARSET_REV = [
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  15,
  -1,
  10,
  17,
  21,
  20,
  26,
  30,
  7,
  5,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  29,
  -1,
  24,
  13,
  25,
  9,
  8,
  23,
  -1,
  18,
  22,
  31,
  27,
  19,
  -1,
  1,
  0,
  3,
  16,
  11,
  28,
  12,
  14,
  6,
  4,
  2,
  -1,
  -1,
  -1,
  -1,
  -1,
  -1,
  29,
  -1,
  24,
  13,
  25,
  9,
  8,
  23,
  -1,
  18,
  22,
  31,
  27,
  19,
  -1,
  1,
  0,
  3,
  16,
  11,
  28,
  12,
  14,
  6,
  4,
  2,
  -1,
  -1,
  -1,
  -1,
  -1
];

int _polymod(Uint8List values) {
  int c = 1;
  values.forEach((v_i) {
    int c0 = (c >> 25) & 0xff;
    c = ((c & 0x1ffffff) << 5) ^ (v_i & 0xff);
    if ((c0 & 1) != 0) c ^= 0x3b6a57b2;
    if ((c0 & 2) != 0) c ^= 0x26508e6d;
    if ((c0 & 4) != 0) c ^= 0x1ea119fa;
    if ((c0 & 8) != 0) c ^= 0x3d4233dd;
    if ((c0 & 16) != 0) c ^= 0x2a1462b3;
  });
  return c;
}

Uint8List _expandHrp(String hrp) {
  var result = hrp.codeUnits.map((c) => c >> 5).toList();
  return Uint8List.fromList(
      result += [0] + hrp.codeUnits.map((c) => c & 31).toList());
}

Uint8List createChecksum(String hrp, Uint8List values) {
  var enc = _expandHrp(hrp) + values + [0, 0, 0, 0, 0, 0];
  int mod = _polymod(Uint8List.fromList(enc)) ^ 1;
  Uint8List ret = Uint8List(6);
  for (int i = 0; i < 6; ++i) {
    ret[i] = (mod >> (5 * (5 - i))) & 31;
  }
  return ret;
}

bool verifyChecksum(String hrp, Uint8List values) {
  return _polymod(Uint8List.fromList(_expandHrp(hrp) + values)) == 1;
}

Uint8List arrayCopy(bytes, srcOffset, result, destOffset, bytesLength) {
  for (var i = srcOffset; i < bytesLength; i++) {
    result[destOffset + i] = bytes[i];
  }
  return result;
}

Uint8List copyOfRange(Uint8List original, int from, int to) {
  int newLength = to - from;
  if (newLength < 0) throw Exception(from.toString() + " > " + to.toString());
  Uint8List copy = Uint8List(newLength);
  return arrayCopy(
      original, from, copy, 0, min(original.length - from, newLength));
}
